# Heads or Tails?

Let's start with a very simple scenario.  We're going to construct a basic set of binary outcomes.  Throughout this multi-week exploration, we'll lean on python as the computing platform and Jupyter notebooks for capturing the workflow.

We'll simulate what happens if the value of something starting at $100$ were to go up or down by $1$ over $5$ days.  Assume the probability of going up or down is 50-50.

In [2]:
import pandas as pd
import numpy as np

In [3]:
starting_value = 100

# List of potential moves in value per day
value_changes = [-1,1]

#This is where we'll store our outcomes.  We'll create it with the initial value
all_outcomes = [[starting_value]]

# Loop through 5 days of changes, and store the sorted outcomes
for day in range(1,6):
    daily_outcomes = []
    for each in value_changes:
        for prior_value in all_outcomes[-1]:
            daily_outcomes.append(prior_value + each)
    all_outcomes.append(np.sort(daily_outcomes))

all_outcomes = all_outcomes[1:]  

In [4]:
for i in range(0,5):
    print('Day {} Outcomes: '.format(i+1), all_outcomes[i])

Day 1 Outcomes:  [ 99 101]
Day 2 Outcomes:  [ 98 100 100 102]
Day 3 Outcomes:  [ 97  99  99  99 101 101 101 103]
Day 4 Outcomes:  [ 96  98  98  98  98 100 100 100 100 100 100 102 102 102 102 104]
Day 5 Outcomes:  [ 95  97  97  97  97  97  99  99  99  99  99  99  99  99  99  99 101 101
 101 101 101 101 101 101 101 101 103 103 103 103 103 105]


In [5]:
df = pd.DataFrame(all_outcomes[4],columns=['Values'])
df.groupby(['Values']).size()

Values
95      1
97      5
99     10
101    10
103     5
105     1
dtype: int64

In [6]:
def build_outcomes(starting_value,value_changes,days):

    #This is where we'll store our outcomes.  We'll create it with the initial value
    all_outcomes = [[starting_value]]

    # Loop through 5 days of changes, and store the sorted outcomes
    for day in range(1,(days+1)):
        daily_outcomes = []
        for each in value_changes:
            for prior_value in all_outcomes[-1]:
                daily_outcomes.append(prior_value + each)
        all_outcomes.append(np.sort(daily_outcomes))

    all_outcomes = all_outcomes[1:] 
    
    return(all_outcomes)

In [7]:
outcomes = build_outcomes(100,[-1,1],10)

In [8]:
for i in range(0,10):
    print('Day {} Outcomes: '.format(i+1), outcomes[i])

Day 1 Outcomes:  [ 99 101]
Day 2 Outcomes:  [ 98 100 100 102]
Day 3 Outcomes:  [ 97  99  99  99 101 101 101 103]
Day 4 Outcomes:  [ 96  98  98  98  98 100 100 100 100 100 100 102 102 102 102 104]
Day 5 Outcomes:  [ 95  97  97  97  97  97  99  99  99  99  99  99  99  99  99  99 101 101
 101 101 101 101 101 101 101 101 103 103 103 103 103 105]
Day 6 Outcomes:  [ 94  96  96  96  96  96  96  98  98  98  98  98  98  98  98  98  98  98
  98  98  98  98 100 100 100 100 100 100 100 100 100 100 100 100 100 100
 100 100 100 100 100 100 102 102 102 102 102 102 102 102 102 102 102 102
 102 102 102 104 104 104 104 104 104 106]
Day 7 Outcomes:  [ 93  95  95  95  95  95  95  95  97  97  97  97  97  97  97  97  97  97
  97  97  97  97  97  97  97  97  97  97  97  99  99  99  99  99  99  99
  99  99  99  99  99  99  99  99  99  99  99  99  99  99  99  99  99  99
  99  99  99  99  99  99  99  99  99  99 101 101 101 101 101 101 101 101
 101 101 101 101 101 101 101 101 101 101 101 101 101 101 101 101 101 

In [12]:
df = pd.DataFrame(outcomes[9],columns=['Values'])
df.groupby(['Values']).size()/1024

Values
90     0.000977
92     0.009766
94     0.043945
96     0.117188
98     0.205078
100    0.246094
102    0.205078
104    0.117188
106    0.043945
108    0.009766
110    0.000977
dtype: float64

Two thirds of the outcomes are within 2 dollars of the starting value.  What would you pay for the option to pay 100 dollars 10 days from now?

In [14]:
2*.2+4*.11+6*0.04+0.01*8

1.1600000000000001

The fair value of that option is worth \$1.16